src/layers/io/input/input_layer.cpp

```diff
@@ -26,9 +26,336 @@
+    auto &device_tensor = this->get_activations(mat_idx);
+
+    // Adjust the mini-batch size
+    original_tensor.set_outermost_dimension(mb_size);
+    host_tensor.set_outermost_dimension(mb_size);
+    device_tensor.set_outermost_dimension(mb_size);
+
+    // Setup view
+    assert0(dc::tensor::View(
+        original_tensor,
+        l.get_activations(mat_idx).LockedBuffer()));
+
+    // Shuffle if necessary
+    if (m_shuffle_required) {
+      get_shuffler(
+          original_tensor, host_tensor, mat_idx).shuffle_forward(
+              original_tensor.get_const_base_ptr(),
+              host_tensor.get_base_ptr());
+    } else {
+      // The input buffer is already partitioned
+      assert0(dc::tensor::View(
+          host_tensor, original_tensor.get_const_buffer()));
+    }
+
+    // After this, there is no inter-process communication, so it's
+    // safe to exit if the local tensor is empty.
+    if (host_tensor.get_local_size() == 0) {
+      continue;
+    }
+
+    prof_region_begin("copy-to-device", prof_colors[1], false);
+    assert0(dc::tensor::Copy(
+        device_tensor, host_tensor, stream));
+    prof_region_end("copy-to-device", false);
+  }
+}
+
+template <typename TensorDataType, typename T_io_buffer,
+          data_layout T_layout, El::Device Dev>
+const input_distconv_adapter<TensorDataType, T_io_buffer, T_layout, Dev>&
+input_layer<TensorDataType, T_io_buffer, T_layout, Dev>::get_distconv_adapter() const {
+  return dynamic_cast<const input_distconv_adapter<
+    TensorDataType, T_io_buffer, T_layout, Dev>&>(
+        data_type_layer<TensorDataType>::get_distconv_adapter());
+}
+
+template <typename TensorDataType, typename T_io_buffer,
+          data_layout T_layout, El::Device Dev>
+input_distconv_adapter<TensorDataType, T_io_buffer, T_layout, Dev>&
+input_layer<TensorDataType, T_io_buffer, T_layout, Dev>::get_distconv_adapter() {
+  return const_cast<input_distconv_adapter<
+    TensorDataType, T_io_buffer, T_layout, Dev>&>(
+        static_cast<const input_layer<
+        TensorDataType, T_io_buffer, T_layout, Dev>&>(*this).get_distconv_adapter());
+}
+
+template <typename TensorDataType,
+          typename T_io_buffer,
+          data_layout T_layout,
+          El::Device Dev>
+bool input_layer<TensorDataType, T_io_buffer, T_layout, Dev>::
+keep_original_outputs(int index) const {
+  if (index == 0) {
+    return data_type_layer<TensorDataType>::keep_original_outputs(index);
+  } else {
+    assert_eq(index, 1);
+    if (this->distconv_enabled() &&
+        get_distconv_adapter().is_input_processed(1)
+        && !get_distconv_adapter().child_copy_required(index)) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+}
+
+template <typename TensorDataType,
+          typename T_io_buffer,
+          data_layout T_layout,
+          El::Device Dev>
+void input_layer<TensorDataType, T_io_buffer, T_layout, Dev>::
+fp_compute() {
+  generic_input_layer<TensorDataType>::fp_compute();
```

Comment:
What happens if `keep_original_outputs` returns false? If we always call `fp_compute`, don't we need to maintain the original buffer for the activations?

Reply:
That's a good question. The original tensor must be always kept, but `keep_original_outputs()` will return `false`. This means `data_type_layer::fp_setup_outputs` is not completed, but as far as I can tell, it still works. I think that that's because the activation matrix is properly resized when `Copy` is called from `partitioned_io_buffer::distribute_from_local_matrix`. 

I'll fix `keep_original_outputs` anway.

Reply:
Fixed with bf1d57492ee5daa4e8295a63d06979a967fc1c76